S3Request.gs

```diff
@@ -135,128 +157,259 @@ S3Request.prototype.execute = function(options) {
-      
       error.toString = function () { return this.message; };
-      
       error.httpRequestLog = this.service.getLastExchangeLog();
     }
-    
     throw error;
   }
-  
+
   return response;
 };
 
+S3Request.prototype.addAuthorization = function(credentials, date) {
+  var datetime = date.toISOString().replace(/[:\-]|\.\d{3}/g, '');
+  if (this.isPresigned()) {
+    this.updateForPresigned(credentials, datetime);
+  } else {
+    this.addHeaders(credentials, datetime);
+  }
+  this.headers['Authorization'] = this.authorization(credentials, datetime)
+}
+
+S3Request.prototype.addHeaders = function (credentials, datetime) {
+  this.headers['X-Amz-Date'] = datetime;
+  if (credentials.sessionToken) {
+    this.headers['x-amz-security-token'] = credentials.sessionToken;
+  }
+}
 
-/* computes Authorization Header value for S3 request
- * reference http://docs.aws.amazon.com/AmazonS3/latest/dev/RESTAuthentication.html
- *
- * @private
- * @return {string} base64 encoded HMAC-SHA1 signature of request (see AWS Rest auth docs for details)
- */
-S3Request.prototype.getAuthHeader_ = function () {
-    
-//  StringToSign = HTTP-VERB + "\n" +
-//    Content-MD5 + "\n" +
-//    Content-Type + "\n" +
-//    Date + "\n" +
-//    CanonicalizedAmzHeaders +
-//    CanonicalizedResource;    
-  var stringToSign = this.httpMethod + "\n";
-  
-  var contentLength = this.content.length;
-  stringToSign += this.getContentMd5_() + "\n" ;
-  stringToSign += this.getContentType() + "\n";
-
-  
-  //set expires time 60 seconds into future
-  stringToSign += this.date.toUTCString() + "\n";
-
-
-  // Construct Canonicalized Amazon Headers
-  //http://docs.aws.amazon.com/AmazonS3/latest/dev/RESTAuthentication.html#RESTAuthenticationRequestCanonicalization
-  var amzHeaders = [];
-  
-  for (var headerName in this.headers) {
-    // only AMZ headers
-    // convert to lower case (1)
-    // multi-line headers to single line (4)
-    // one space after : (5)
-    if (headerName.match(/^x-amz/i)) {
-      var header = headerName.toLowerCase() + ":" + this.headers[headerName].replace(/\s+/, " ");
-      amzHeaders.push(header) 
+S3Request.prototype.updateForPresigned = function(credentials, datetime) {
+  var credString = this.credentialString(datetime);
+  var qs = {
+    'X-Amz-Date': datetime,
+    'X-Amz-Algorithm': 'AWS4-HMAC-SHA256',
+    'X-Amz-Credential': credentials.accessKeyId + '/' + credString,
+    'X-Amz-Expires': this.headers[this.expiresHeader],
+    'X-Amz-SignedHeaders': this.signedHeaders()
+  };
+
+  if (credentials.sessionToken) {
+    qs['X-Amz-Security-Token'] = credentials.sessionToken;
+  }
+
+  if (this.headers['Content-Type']) {
+    qs['Content-Type'] = this.headers['Content-Type'];
+  }
+  if (this.headers['Content-MD5']) {
+    qs['Content-MD5'] = this.headers['Content-MD5'];
+  }
+  if (this.headers['Cache-Control']) {
+    qs['Cache-Control'] = this.headers['Cache-Control'];
+  }
```

Comment:
seems weird to not have these cases covered by your loop below.  In essence, you're including all of these in the qs to be signed. so why not have `isSignableHeader()` return true for them? eg, you have `content-type` as something not to be signed in the body of `isSignableHeader()`

Reply:
I have not enough understanding about these part. But I ported the master(AWS official) code [v4.js](https://github.com/aws/aws-sdk-js/blob/master/lib/signers/v4.js#L65-L67) 
I think v4.js is a common library, and other service (not s3) requires these spec. (ex. Cloud Front processes the Cache-Control header and do not pass it to S3)

Reply:
yeah, I agree that AMZN's version of the lib has this structure problem too. It makes it hard to follow, and therefore hard to you to port over the apps script.

Reply:
yeah,  I had tried to transplant the whole thing but gave up...